sticker-utils/src/subcommands/pretrain.rs

```diff
@@ -122,161 +120,180 @@ impl PretrainApp {
+        let train_progress = ProgressBar::new(n_steps as u64);
+        train_progress.set_style(ProgressStyle::default_bar().template(
+            "[Time: {elapsed_precise}, ETA: {eta_precise}] {bar} {percent}% train {msg}",
+        ));
 
-            let (loss, acc, _) = self.run_epoch(
-                &mut saver,
+        // Continue performing steps until we are done.
+        while global_step < n_steps - 1 {
+            let mut train_dataset = Self::open_dataset(&train_file)?;
+            let train_batches = train_dataset.batches(
                 &categorical_encoder,
                 &vectorizer,
-                &trainer,
-                &mut validation_file,
-                false,
-                train_size,
-                global_step,
-                epoch,
+                self.batch_size,
+                self.max_len,
+                self.shuffle_buffer_size,
             )?;
 
-            saver
-                .save(&trainer, CompletedUnit::Epoch(acc))
-                .or_exit("Error saving model", 1);
+            for steps in &train_batches.chunks(self.eval_steps) {
+                let (_, _, global_step_after_steps) =
+                    self.train_steps(&train_progress, &trainer, steps, global_step, n_steps)?;
+                global_step = global_step_after_steps;
 
-            if acc > best_acc {
-                best_epoch = epoch;
-                best_acc = acc;
-            }
+                let (loss, acc) = self.validation_epoch(
+                    &trainer,
+                    &validation_file,
+                    &categorical_encoder,
+                    &vectorizer,
+                    global_step,
+                )?;
 
-            let epoch_status = if best_epoch == epoch { "🎉" } else { "" };
+                if acc > best_acc {
+                    best_step = global_step;
+                    best_acc = acc;
 
-            eprintln!(
-            "Epoch {} (validation): loss: {:.4}, acc: {:.4}, best epoch: {}, best acc: {:.4} {}",
-            epoch, loss, acc, best_epoch, best_acc, epoch_status
-        );
+                    trainer
+                        .save(format!("{}step-{}", "pretrain-", global_step))
+                        .context(format!("Cannot save model for step {}", global_step))?;
+                }
+
+                let step_status = if best_step == global_step { "🎉" } else { "" };
+
+                eprintln!("Step {} (validation): loss: {:.4}, acc: {:.4}, best step: {}, best acc: {:.4} {}\n",
+			  global_step, loss, acc, best_step, best_acc, step_status);
+
+                if global_step >= n_steps - 1 {
+                    break;
+                }
+            }
         }
 
         Ok(())
     }
 
     #[allow(clippy::too_many_arguments)]
-    fn run_epoch<E>(
+    fn train_steps(
+        &self,
+        progress: &ProgressBar,
+        trainer: &TaggerTrainer,
+        steps: impl Iterator<Item = Fallible<TensorBuilder<LabelTensor>>>,
```

Comment:
rename `steps` to `batches` or something more descriptive. With `steps` I'd expect it to define how many steps this function call should perform.

Reply:
Fixed.